### PR TITLE
ARROW-2784: [C++] MemoryMappedFile::WriteAt allow writing past the end

### PR DESCRIPTION
**Pull request by alendit**

### CODE REVIEW

File: cpp/src/arrow/io/file.cc [hunk 1]
```diff
@@ -517,6 +517,10 @@ Status MemoryMappedFile::WriteAt(int64_t position, const void* data, int64_t nby
   }
 
   RETURN_NOT_OK(memory_map_->Seek(position));
+  if (nbytes + memory_map_->position() > memory_map_->size()) {
```
Comment (pitrou) [inline cpp/src/arrow/io/file.cc]: We should probably do this check before seeking.

**Review by pitrou — CHANGES_REQUESTED** [review]
Pending the CI results, I have one comment below.

File: cpp/src/arrow/io/file.cc (hunk 1 above)
Comment (alendit) [reply cpp/src/arrow/io/file.cc]: Absolutely. I added a check for it.

**Comment by alendit** [comment]
I also added a pure `WriteAt` test, since there wasn't one.

File: cpp/src/arrow/io/io-file-test.cc [hunk 2]
```diff
@@ -614,6 +614,66 @@ TEST_F(TestMemoryMappedFile, WriteRead) {
...
+  test::random_bytes(buffer_size, 0, buffer.data());
+
+  std::string path = "io-memory-map-write-read-test";
+  std::shared_ptr<MemoryMappedFile> result;
+  ASSERT_OK(InitMemoryMap(buffer_size, path, &result));
+
+  int64_t position;
+  ASSERT_OK(result->WriteAt(0, buffer.data(), buffer_size / 2));
+  ASSERT_OK(result->Tell(&position));
+  ASSERT_EQ(position, buffer_size / 2);
```
Comment (pitrou) [inline cpp/src/arrow/io/io-file-test.cc]: Interesting. I wouldn't expect `WriteAt` to change the internal position at all. I'd rather remove those lines from the test (and perhaps `WriteAt` to remove the seeking, which hopefully won't break anything :-)).

Comment (alendit) [reply cpp/src/arrow/io/io-file-test.cc]: It doesn't break any tests, at least. But `FixedSizeBufferWriter` and `CudaMemory` implement `WriteAt` in such a way, that it moves the position. 

I've pushed the commit changing `WriteAt` behaviour for the `MemoryMappedFile`, but I'd rather revert it: I think a bit surprising, but consistent behaviour is preferable here.

Comment (pitrou) [reply cpp/src/arrow/io/io-file-test.cc]: Ok, I guess we can fix that in another PR. But I'd rather not have the tests ascertain the current behaviour :-)

Comment (alendit) [reply cpp/src/arrow/io/io-file-test.cc]: Removed tests, kept behaviour for now.

**Review by wesm — APPROVED** [review]
+1, thank you @alendit!

**Pull request closed without merging**